### PR TITLE
Fix Service.free logic

### PR DESCRIPTION
**Pull request by granders**
@ewencp 
Fairly small PR here. v0.3.3 had a dumb error resulting in the cluster gradually running out of nodes because nodes were being improperly freed by service classes with multiple nodes.

Fixes https://github.com/confluentinc/ducktape/issues/103

### CODE REVIEW

File: ducktape/services/service.py [hunk 1]
```diff
@@ -92,13 +92,14 @@ def allocate_nodes(self):
         if self.allocated:
             raise Exception("Requesting nodes for a service that has already been allocated nodes.")
 
-        self.logger.debug("Requesting nodes from the cluster.")
+        self.logger.debug("Requesting %d nodes from the cluster." % self.num_nodes)
 
         try:
             self.nodes = self.cluster.request(self.num_nodes)
         except RuntimeError as e:
+            msg = ""
```
Comment (ewencp) [inline ducktape/services/service.py]: Shouldn't the `str(e.message)` be outside the if statement?

File: ducktape/services/service.py [hunk 2]
```diff
@@ -181,10 +184,11 @@ def free(self):
             node.account.logger = None
             self.free_node(node)
 
+        self.nodes = []
+
     def free_node(self, node):
```
Comment (ewencp) [inline ducktape/services/service.py]: Can we get rid of `free_node`? It's not safe to call on its own -- it looks like `free` is supposed to be the only safe way to clean up nodes now. At a minimum, seems like this should be `_free_node`.

File: ducktape/tests/reporter.py [hunk 3]
```diff
@@ -63,7 +63,7 @@ def result_string(self):
             result_lines.append("    " + self.result.summary)
 
         if self.result.data is not None:
-            result_lines.append(json.dumps(self.result.data))
+            result_lines.append(pprint.pformat(self.result.data))
```
Comment (ewencp) [inline ducktape/tests/reporter.py]: These changes seem unrelated to the rest of the patch. Why change to pprint? In the parts below where it writes to `.json` files, is the output guaranteed to be JSON compatible? It seems like the places where we cared about pretty printing we were already properly indenting the JSON.

**Comment by ewencp** [comment]
Just so I'm sure I understand the issue correctly, the problem was that there was a call to `list.remove` while iterating over the `list` of nodes?

Left a few comments, but looks good. Feel free to merge after addressing comments (and deciding whether to include those `pprint` changes).

File: ducktape/services/service.py (hunk 1 above)
Comment (granders) [inline ducktape/services/service.py]: yep, done

File: ducktape/services/service.py (hunk 2 above)
Comment (granders) [inline ducktape/services/service.py]: done

File: ducktape/tests/reporter.py (hunk 3 above)
Comment (granders) [inline ducktape/tests/reporter.py]: Yeah, this was a bit iffy.
I wanted to support the use case where functions are used to parameterize tests, e.g. `@parameterize(x=my_function)`, and `json.dumps` doesn't know how to stringify functions.

This is also tricky if we also want to be able to specify which parameters to use in a test at the command-line.

rolling this back

**Comment by granders** [comment]
@ewencp 
Yes, that right, the problem was calling `list.remove` while iterating over the list.
